Generator/generator-botbuilder-java/generators/app/index.js

```diff
@@ -1,49 +1,267 @@
 'use strict';
+const pkg = require('../../package.json');
 const Generator = require('yeoman-generator');
-const chalk = require('chalk');
-const yosay = require('yosay');
 const path = require('path');
 const _ = require('lodash');
-const extend = require('deep-extend');
+const chalk = require('chalk');
 const mkdirp = require('mkdirp');
 
+const BOT_TEMPLATE_NAME_EMPTY = 'Empty Bot';
+const BOT_TEMPLATE_NAME_SIMPLE = 'Echo Bot';
+const BOT_TEMPLATE_NAME_CORE = 'Core Bot';
+
+const BOT_TEMPLATE_NOPROMPT_EMPTY = 'empty';
+const BOT_TEMPLATE_NOPROMPT_SIMPLE = 'echo';
+const BOT_TEMPLATE_NOPROMPT_CORE = 'core';
+
 module.exports = class extends Generator {
-  prompting() {
-    // Have Yeoman greet the user.
-    this.log(
-      yosay(`Welcome to the badass ${chalk.red('generator-botbuilder-java')} generator!`)
-    );
-
-    const prompts = [
-      { name: 'botName', message: `What 's the name of your bot?`, default: 'sample' },
-      { name: 'description', message: 'What will your bot do?', default: 'sample' },
-      { name: 'dialog', type: 'list', message: 'What default dialog do you want?', choices: ['Echo'] },
-    ];
-
-    return this.prompt(prompts).then(props => {
-      // To access props later use this.props.someAnswer;
-      this.props = props;
-    });
-  }
-
-  writing() {
-    const directoryName = _.kebabCase(this.props.botName);
-    const defaultDialog = this.props.dialog.split(' ')[0].toLowerCase();
-
-    if (path.basename(this.destinationPath()) !== directoryName) {
-      this.log(`Your bot should be in a directory named ${directoryName}\nI'll automatically create this folder.`);
-      mkdirp(directoryName);
-      this.destinationRoot(this.destinationPath(directoryName));
-    }
-
-    this.fs.copyTpl(this.templatePath('pom.xml'), this.destinationPath('pom.xml'), { botName: directoryName });
-    this.fs.copy(this.templatePath(`app.java`), this.destinationPath(`app.java`));
-    this.fs.copyTpl(this.templatePath('README.md'), this.destinationPath('README.md'), {
-      botName: this.props.botName, description: this.props.description
-    });
-  }
-
-  install() {
-    this.installDependencies({bower: false});
-  }
+    constructor(args, opts) {
+        super(args, opts);
+
+        // allocate an object that we can use to store our user prompt values from our askFor* functions
+        this.templateConfig = {}
+
+        // configure the commandline options
+        this._configureCommandlineOptions();
+    }
+
+    initializing() {
+        // give the user some data before we start asking them questions
+        this.log(`\nWelcome to the Microsoft Java Bot Builder generator v${pkg.version}. `);
+    }
+
+    prompting() {
+        const userPrompts = this._getPrompts();
+
+        // if we're told to not prompt, then pick what we need and return
+        if(this.options.noprompt) {
+            // this function will throw if it encounters errors/invalid options
+            this._verifyNoPromptOptions();
+        }
```

Comment:
If this logic occurs before `const userPrompts = this._getPrompts();` on Line 37 is called, you won't need the `if(noprompt) {` logic in `_getPrompts()`

Since `_verifyNoPromptOptions()` throws an Error or returns `undefined`, and the prompts modify `this.templateConfig`, the refactoring could probably look like this:

```js
prompting() {
  // if we're told to not prompt, then check for required parameters
  if(this.options.noprompt) {
    // this function will throw if it encounters errors/invalid options
    return this._verifyNoPromptOptions();
  }

  const userPrompts = this._getPrompts();
  async function executePrompts([prompt, ...rest]) {
    if (prompt) {
        await prompt();
        return executePrompts(rest);
      }
   }

  return executePrompts(userPrompts);
}
```

Granted this means that the caller of `prompting()` doesn't require a specific return value.

Reply:
Return Promise.resolve() in that case?